apiserver/facades/agent/caasoperator/watcher.go

```diff
@@ -0,0 +1,109 @@
+// Copyright 2019 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package caasoperator
+
+import (
+	"github.com/juju/errors"
+	"github.com/juju/utils/set"
+	"gopkg.in/juju/worker.v1"
+	"gopkg.in/juju/worker.v1/catacomb"
+
+	corewatcher "github.com/juju/juju/core/watcher"
+	"github.com/juju/juju/state"
+)
+
+type unitIDWatcher struct {
+	catacomb catacomb.Catacomb
+	out      chan []string
+	src      corewatcher.StringsWatcher
+	model    watcherModelInterface
+}
+
+type watcherModelInterface interface {
+	Containers(providerIds ...string) ([]state.CloudContainer, error)
+}
+
+// newUnitIDWatcher watches a StringsWatcher and converts ProviderIDs into UnitIDs and re-emits them.
+func newUnitIDWatcher(model watcherModelInterface, src corewatcher.StringsWatcher) (*unitIDWatcher, error) {
+	w := &unitIDWatcher{
+		out:   make(chan []string),
+		src:   src,
+		model: model,
+	}
+	err := catacomb.Invoke(catacomb.Plan{
+		Site: &w.catacomb,
+		Work: w.loop,
+		Init: []worker.Worker{src},
+	})
+	return w, err
+}
+
+func (w *unitIDWatcher) Kill() {
+	w.catacomb.Kill(nil)
+}
+
+func (w *unitIDWatcher) Wait() error {
+	return w.catacomb.Wait()
+}
+
+func (w *unitIDWatcher) Stop() error {
+	w.Kill()
+	return w.Wait()
+}
+
+func (w *unitIDWatcher) Changes() <-chan []string {
+	return w.out
+}
+
+func (w *unitIDWatcher) Err() error {
+	return w.catacomb.Err()
+}
+
+func (w *unitIDWatcher) loop() error {
+	defer close(w.out)
+
+	var out chan []string
+	var result []string
+
+	// initial event is sent regardless of if it's
+	// empty.
+	sendInitial := true
+
+	for {
+		select {
+		case <-w.catacomb.Dying():
+			return w.catacomb.ErrDying()
+		case providerIDs, ok := <-w.src.Changes():
+			if !ok {
+				return errors.Errorf("event watcher closed")
+			}
+			if len(providerIDs) == 0 {
+				if sendInitial {
+					out = w.out
+					sendInitial = false
```

Comment:
don't set to false here (and a second time below) - set to false in the out <- results bit